internal/fwserver/server_listresources.go

```diff
@@ -0,0 +1,135 @@
+
+	logging.FrameworkTrace(ctx, "Calling provider defined ListResources")
+	listResourceFuncSlice := provider.ListResources(ctx)
+	logging.FrameworkTrace(ctx, "Called provider defined ListResources")
+
+	for _, listResourceFunc := range listResourceFuncSlice {
+		listResource := listResourceFunc()
+
+		metadataReq := resource.MetadataRequest{
+			ProviderTypeName: providerTypeName,
+		}
+		metadataResp := resource.MetadataResponse{}
+		listResource.Metadata(ctx, metadataReq, &metadataResp)
+
+		typeName := metadataResp.TypeName
+		if typeName == "" {
+			s.listResourceFuncsDiags.AddError(
+				"ListResource Type Name Missing",
+				fmt.Sprintf("The %T ListResource returned an empty string from the Metadata method. ", listResource)+
+					"This is always an issue with the provider and should be reported to the provider developers.",
+			)
+			continue
+		}
+
+		logging.FrameworkTrace(ctx, "Found resource type", map[string]interface{}{logging.KeyListResourceType: typeName}) // TODO: y?
+
+		if _, ok := s.listResourceFuncs[typeName]; ok {
+			s.listResourceFuncsDiags.AddError(
+				"Duplicate ListResource Type Defined",
+				fmt.Sprintf("The %s ListResource type name was returned for multiple list resources. ", typeName)+
+					"ListResource type names must be unique. "+
+					"This is always an issue with the provider and should be reported to the provider developers.",
+			)
+			continue
+		}
+
+		if _, ok := s.resourceFuncs[typeName]; !ok {
+			s.listResourceFuncsDiags.AddError(
+				"ListResource Type Defined without a Matching Managed Resource Type",
+				fmt.Sprintf("The %s ListResource type name was returned, but no matching managed Resource type was defined. ", typeName)+
+					"This is always an issue with the provider and should be reported to the provider developers.",
+			)
+			continue
+		}
+
+		s.listResourceFuncs[typeName] = listResourceFunc
+	}
+
+	return s.listResourceFuncs, s.listResourceFuncsDiags
+}
+
+// ListResourceMetadatas returns a slice of ListResourceMetadata for the GetMetadata
+// RPC.
+func (s *Server) ListResourceMetadatas(ctx context.Context) ([]ListResourceMetadata, diag.Diagnostics) {
+	resourceFuncs, diags := s.ListResourceFuncs(ctx)
+
+	resourceMetadatas := make([]ListResourceMetadata, 0, len(resourceFuncs))
+
+	for typeName := range resourceFuncs {
+		resourceMetadatas = append(resourceMetadatas, ListResourceMetadata{
+			TypeName: typeName,
+		})
+	}
+
+	return resourceMetadatas, diags
+}
+
+// ListResourceSchemas returns a map of ListResource Schemas for the
+// GetProviderSchema RPC without caching since not all schemas are guaranteed to
+// be necessary for later provider operations. The schema implementations are
+// also validated.
+func (s *Server) ListResourceSchemas(ctx context.Context) (map[string]fwschema.Schema, diag.Diagnostics) {
+	listResourceSchemas := make(map[string]fwschema.Schema)
+	listResourceFuncs, diags := s.ListResourceFuncs(ctx)
```

Comment:
This will give us the list resource implementations actually defined in `ListResources()`, but couldn't there also be implementations in `Resources()` as well? Potentially conflicting 👀 

> same thought for metadata

Reply:
My current thinking on this is: Framework only needs to look at `ListResources()`.

If a provider type implements both `resource.Resource` and `list.ListResource`, then they add a `NewMyResource` function in two places: `Resources()` and `ListResources()`.

Example: https://github.com/hashicorp/terraform-provider-corner/pull/358/files#diff-8e71dd37ab963ab4edd62bcbc12cc75b0d4cee4e084f515bf239385c4d16b7e2

There's room for improvement...